src/nxt_otel.c

```diff
@@ -0,0 +1,403 @@
+        traceparent_name = (nxt_str_t){
+            .start  = f->name,
+            .length = f->name_length,
+        };
+        traceparent = (nxt_str_t){
+            .start  = f->value,
+            .length = f->value_length,
+        };
+        nxt_otel_rs_add_event_to_trace(r->otel->trace,
+                                       &traceparent_name,
+                                       &traceparent);
+
+    // potentially nxt_http_request_error called before headers finished parsing
+    } else {
+        nxt_log(task, NXT_LOG_DEBUG,
+                "not propagating tracing headers for missing trace");
+        return;
+    }
+
+    f = nxt_list_add(r->resp.fields);
+    if (f == NULL) {
+        nxt_log(task, NXT_LOG_ERR,
+                "couldnt allocate traceparent header in response");
+        return;
+    }
+
+    nxt_http_field_name_set(f, "traceparent");
+    f->value = traceval;
+    f->value_length = nxt_strlen(traceval);
+}
+
+
+void
+nxt_otel_span_add_headers(nxt_task_t *task, nxt_http_request_t *r)
+{
+    nxt_str_t        method_name, path_name;
+    nxt_http_field_t *cur;
+
+    nxt_log(task, NXT_LOG_DEBUG, "adding headers to trace");
+
+    if (r->otel == NULL || r->otel->trace == NULL)
+    {
+        nxt_log(task, NXT_LOG_ERR, "no trace to add events to!");
+        nxt_otel_state_transition(r->otel, NXT_OTEL_ERROR_STATE);
+        return;
+    }
+
+    nxt_list_each(cur, r->fields) {
+        nxt_str_t name, val;
+        name = (nxt_str_t){
+            .start  = cur->name,
+            .length = cur->name_length,
+        };
+
+        val = (nxt_str_t){
+            .start  = cur->value,
+            .length = cur->value_length,
+        };
+
+        nxt_otel_rs_add_event_to_trace(r->otel->trace, &name, &val);
+    } nxt_list_loop;
+
+    method_name = (nxt_str_t){
+        .start  = (u_char *) NXT_OTEL_METHOD_TAG,
+        .length = nxt_length(NXT_OTEL_METHOD_TAG) + 1,
+    };
+    nxt_otel_rs_add_event_to_trace(r->otel->trace, &method_name, r->method);
+
+    path_name = (nxt_str_t){
+        .start  = (u_char *) NXT_OTEL_PATH_TAG,
+        .length = nxt_length(NXT_OTEL_PATH_TAG) + 1,
+    };
+    nxt_otel_rs_add_event_to_trace(r->otel->trace, &path_name, r->path);
+    nxt_otel_propagate_header(task, r);
+
+    nxt_otel_state_transition(r->otel, NXT_OTEL_BODY_STATE);
+}
+
+
+void
+nxt_otel_span_add_body(nxt_http_request_t *r)
+{
+    size_t    body_size, buf_size;
+    u_char    *body_buf, *body_size_buf;
+    nxt_str_t body_key, body_val;
+    nxt_int_t cur;
+
+    if (r->body != NULL) {
+        body_size = nxt_buf_used_size(r->body);
+    } else {
+        body_size = 0;
+    }
+
+    buf_size = 1; // first digit
+    if (body_size != 0) {
+        buf_size += log10(body_size); // subsequent digits
+    }
+    buf_size += 1; // \0
+    buf_size += nxt_length(NXT_OTEL_BODY_SIZE_TAG);
+    buf_size += 1; // \0
+
+    body_buf = nxt_mp_zalloc(r->mem_pool, buf_size);
```

Comment:
because ..... ?

Reply:
sprintf(3) will add a `\0` and having just double checked exactly what nxt_cpystr() does, it will copy the `\0` from the source, so both of these give you nul-terminated strings.

You can then get rid of those comments (one of which is sort of wrong anyway).